core/lib/Thelia/Action/ProductSaleElement.php

```diff
@@ -299,6 +309,137 @@ protected function createCombination(ConnectionInterface $con, ProductSaleElemen
+            ->findOneByProductSaleElementsId($originalProductPSE->getId());
+
+        $clonedProductUpdatePSEEvent = new ProductSaleElementUpdateEvent($clonedProduct, $clonedProductPSEId);
+        $clonedProductUpdatePSEEvent
+            ->setReference($clonedProduct->getRef().'-'.($key + 1))
+            ->setIsdefault($originalProductPSE->getIsDefault())
+            ->setFromDefaultCurrency(0)
+
+            ->setWeight($originalProductPSE->getWeight())
+            ->setQuantity($originalProductPSE->getQuantity())
+            ->setOnsale($originalProductPSE->getPromo())
+            ->setIsnew($originalProductPSE->getNewness())
+            ->setEanCode($originalProductPSE->getEanCode())
+            ->setTaxRuleId($originalProduct->getTaxRuleId())
+
+            ->setPrice($originalProductPSEPrice->getPrice())
+            ->setSalePrice($originalProductPSEPrice->getPromoPrice())
+            ->setCurrencyId($originalProductPSEPrice->getCurrencyId());
+
+        $dispatcher->dispatch(TheliaEvents::PRODUCT_UPDATE_PRODUCT_SALE_ELEMENT, $clonedProductUpdatePSEEvent);
+    }
+
+    public function clonePSEAssociatedImages($clonedProductId, $clonedProductPSEId, $originalProductPSEImages)
+    {
+        foreach ($originalProductPSEImages as $originalProductPSEImage) {
+            $originalProductImagePosition = ProductImageQuery::create()
+                ->select(['POSITION'])
+                ->findPk($originalProductPSEImage->getProductImageId());
+
+            $clonedProductImageIdToLinkToPSE = ProductImageQuery::create()
+                ->filterByProductId($clonedProductId)
+                ->filterByPosition($originalProductImagePosition)
+                ->select(['ID'])
+                ->findOne();
+
+            $assoc = new ProductSaleElementsProductImage();
+            $assoc
+                ->setProductSaleElementsId($clonedProductPSEId)
+                ->setProductImageId($clonedProductImageIdToLinkToPSE)
+                ->save();
+        }
+    }
+
+    public function clonePSEAssociatedDocuments($clonedProductId, $clonedProductPSEId, $originalProductPSEDocuments)
+    {
+        foreach ($originalProductPSEDocuments as $originalProductPSEDocument) {
+            $originalProductDocumentPosition = ProductDocumentQuery::create()
+                ->select(['POSITION'])
+                ->findPk($originalProductPSEDocument->getProductDocumentId());
+
+            $clonedProductDocumentIdToLinkToPSE = ProductDocumentQuery::create()
+                ->filterByProductId($clonedProductId)
+                ->filterByPosition($originalProductDocumentPosition)
+                ->select(['ID'])
+                ->findOne();
+
+            $assoc = new ProductSaleElementsProductDocument();
+            $assoc
+                ->setProductSaleElementsId($clonedProductPSEId)
+                ->setProductDocumentId($clonedProductDocumentIdToLinkToPSE)
+                ->save();
+        }
+    }
+
```

Comment:
clonePSEAssociatedImages() and clonePSEAssociatedDocuments() are mostrly identical, except the Product<obj>Query and ProductSaleElementsProduct<obj> classes names.

You may try to factorize them (https://pbs.twimg.com/media/CF7DsDDWAAAOrnG.jpg)